README.md
# Jeti_AmpSensor
Jeti AmpSensor

Code for Jeti AmpSensor

 Supported sensor modules:
  - ACS712 - 5/20/30A versions
  - ACS758 - 50/100/150/200 B/U versions

 Generates four values to transmitter:
  - Voltage (V) Max. 60V
  - Current (A) Max. 200A cont.
  - Capacity (mAh)
  - Power (W)

Possibility to store capacity between flights, resettable
with digital output from receiver.

User calibratable for current measurement.

More info: https://www.rc-thoughts.com/jeti-ampsensor/

Shared under MIT-license by Tero Salminen (c) RC-Thoughts 2017-2021


CHANGELOG.txt
------------------------------------------------------------------
                   RCT Amp Sensor
------------------------------------------------------------------

 Tero Salminen RC-Thoughts.com (c) 2017-2021 www.rc-thoughts.com

------------------------------------------------------------------

2017-12-27  Tero Salminen  <tero@rc-thoughts.com>

    * v.1.0 First release
    
2018-07-19  Tero Salminen <tero@rc-thoughts.com>

    * v.1.2 Improved small current measurement and accuracy

2021-03-27

    * v.1.4 Corrected saved Capacity, added LED to indicate stored capacity

Arduino/RCT-AmpSensor/Jeti_Init.h
//
// Stuff done in initialization of sketch
//

// Jetibox start screen
#define ABOUT_1 F(" RCT Jeti Tools")
#define ABOUT_2 F(" AmpSensor 1.4")

// Max screens in Jetibox
#define MAX_SCREEN 10

// Sensor type selection
enum sensorType {
  None,
  ACS712_05A,
  ACS712_20A,
  ACS712_30A,
  ACS758_50B,
  ACS758_100B,
  ACS758_150B,
  ACS758_200B,
  ACS758_50U,
  ACS758_100U,
  ACS758_150U,
  ACS758_200U
};

// Offset for sensors
const int ACSoffsetArray[] =  {
  2500,   // None
  2500,   // ACS712-05
  2500,   // ACS712-20
  2500,   // ACS712-30
  2500,   // ACS758-50B
  2500,   // ACS758-100B
  2500,   // ACS758-150B
  2500,   // ACS758-200B
  600,    // ACS758-50U
  600,    // ACS758-100U
  600,    // ACS758-150U
  600     // ACS758-200U
};

// Sensors mV per Amp ratio
const float mVperAmpArray[] =  {
  100.0,    // None
  185.0,    // ACS712-05
  100.0,    // ACS712-20
  66.0,     // ACS712-30
  40.0,     // ACS758-50B
  20.0,     // ACS758-100B
  13.3,     // ACS758-150B
  10.0,     // ACS758-200B
  60.0,     // ACS758-50U
  40.0,     // ACS758-100U
  26.7,     // ACS758-150U
  20.0      // ACS758-200U
};

// Voltage Divider selection
enum dividerType {
  Max_14S_60V,
  Max_6S_30V,
  Max_3S_15V,
  Rx_10V
};

// Voltage Divider R1
const int R1[] = {
  11000,
  10000,
  2000,
  2000
};

// Voltage Divider R2
const int R2[] = {
  1000,
  2000,
  1000,
  2000
};

// ID number for sensor (Needs to be unique for all sensors in model 0xA400 – 0xA41F)
const unsigned char Jeti_SensorID4 = 0x09;

//
// Sensor variables & other init-stuff needed
//

// Input pins
const int analogInCur = A3;
const int analogInVolt = A2;

// Jeti values
float uCurrent = 0;
float uCapacity = 0; // was int
float uVoltage = 0;
long uPower = 0;
long jetiCurrent = 0;
long jetiVoltage = 0;
long jetiCapacity = 0;

// Voltage divider - Defaults to 14S / 60V
int dividerR1;
int dividerR2;
float divider;

// Calibration & calculation values
float mVperAmp = 0;
int ACSoffset = 0;
int curCalVal = 0;
int tempcurCalVal = 100;
int voltCalVal = 0;
int tempvoltCalVal = 100;
static float tmpCap = 0;
unsigned long prevMillis = 0; // Previous time in ms
unsigned long curMillis = 0;  // Current time in ms

// Defaults
int sensorType = 0;
int voltageRange = 0;

// Reset function
int pinReset = 5;
int resetFunction = 0;
int resetState = 0;
int currentActive = 0;
int currentStored = 0;
long resetTime = 0;
long curTime = 0;


Arduino/RCT-AmpSensor/Jeti_JB_Navi.h
//
// Jetibox navigation and button actions
//
/*
  Button codes
  224  RIGHT
  112  LEFT
  208  UP
  176  DOWN
  144  UP & DOWN
  9    LEFT & RIGHT
*/
case 224 : // RIGHT
if (current_screen != MAX_SCREEN)
{
  if (current_screen == 9) {
    EEPROM.write(0, 1); // Reset byte set to done
    EEPROM.write(1, resetFunction); // Reset-function, not in use = 0, in use = 1
    EEPROM.write(2, sensorType); // Sensor type, default is Not Set
    EEPROM.write(3, voltageRange); // Voltage range, default is 14S 60V
    EEPROM.write(4, curCalVal); // Calibration value for current, default is 100
    EEPROM.put(10, voltCalVal); // Calibration value for voltage, default is 100
    EEPROM.put(20, 0);  // Saved capacity (mAh)
    delay(200);
    resetSensor();
  } else {
    current_screen++;
  }
}
break;

case 112 : // LEFT
if (current_screen == 99) {
  current_screen = 0;
}
if (current_screen != 0) {
  current_screen--;
}
break;

case 208 : // UP
if (current_screen == 3) {
  if (sensorType < 12) sensorType++;
  if (sensorType == 12) sensorType = 0;
}
if (current_screen == 4) {
  if (voltageRange < 4) voltageRange++;
  if (voltageRange == 4) voltageRange = 0;
}
if (current_screen == 5) {
  curCalVal++;
}
if (current_screen == 6) {
  voltCalVal++;
}
if (current_screen == 7) {
  if (resetFunction == 0) {
    resetFunction = 1;
  } else {
    resetFunction = 0;
  }
}
break;

case 176 : // DOWN
if (current_screen == 3) {
  if (sensorType == 0) sensorType = 12;
  if (sensorType > 0) sensorType--;
}
if (current_screen == 4) {
  if (voltageRange < 4) voltageRange++;
  if (voltageRange == 4) voltageRange = 0;
}
if (current_screen == 5) {
  curCalVal--;
}
if (current_screen == 6) {
  voltCalVal--;
}
if (current_screen == 7) {
  if (resetFunction == 0) {
    resetFunction = 1;
  } else {
    resetFunction = 0;
  }
}
break;

case 144 : // UP+DOWN
if (current_screen == 8) {
  EEPROM.write(0, 0); // Reset byte
  delay(200);
  resetSensor();
}
break;

case 96 : // LEFT+RIGHT
break;


Arduino/RCT-AmpSensor/Jeti_Runtime.h
//
// This is the main loop where the work is done
// May the current be with you
//

if (sensorType != 0) {
  // Calculate voltage (V)
  uVoltage = (analogRead(analogInVolt) / 1023.0) * (5010 / 1000) / divider * (voltCalVal / 1000.0);

  // Calculate current (A)
  float mVoltCur = (analogRead(analogInCur) / 1023.0) * 5010; // mV
  uCurrent = (mVoltCur - ACSoffset + tempcurCalVal) / mVperAmp;
  if (uCurrent < 0) {
    uCurrent = 0;
  }

  // Calculate power (W)
  uPower = uCurrent * uVoltage;

  // mAh calculation 2 times a second
  if (millis() > prevMillis + 500) {
    tmpCap = 0;
    curMillis = millis() - prevMillis;
    tmpCap = tmpCap + (uCurrent * 1000.0) * ((float)curMillis / 3600000.0);
    uCapacity = uCapacity + tmpCap;
    prevMillis = millis();
  }

  // If reset-function is in use set timer if current is over 1A
  if (resetFunction == 1 && uCurrent > 1) {
    currentActive = 1;
    currentStored = 0;
    resetTime = millis() + 15000;
    digitalWrite(13, LOW);
  } else {
    currentActive = 0;
  }

  // If reset-function is in use and if no current consumed for 15 seconds store consumed capacity once to EEPROM
  curTime = millis();
  if (resetFunction == 1 && currentActive == 0 && currentStored == 0 && curTime > resetTime && ((int)uCapacity > 0)) {
    writeEEPROM(20, (int)uCapacity);
    currentStored = 1;
    digitalWrite(13, HIGH);
  }

  // If reset-function is in use reset consumed capacity from Digital Output
  // allowed only if there is no current used! (Motor off)
  if (resetFunction == 1) {
    resetState = digitalRead(pinReset); // HIGH when open, LOW when closed to GND
    if (currentActive == 0 && resetState == LOW) {
      writeEEPROM(20, 0);
      currentStored = 1;
      uCapacity = 0;
      digitalWrite(13, LOW);
    }
  }
}
else {
  uVoltage = 0;
  uCurrent = 0;
  uPower = 0;
  uCapacity = 0;
}
jetiVoltage = uVoltage * 100;
jetiCurrent = uCurrent * 10;
jetiCapacity = (long)uCapacity;


Arduino/RCT-AmpSensor/Jeti_JB_Screens.h
//
// Build Jetibox screens
//

case 0 : { // aboutScreen
  JB.JetiBox_P(ABOUT_1, ABOUT_2);
  break;
}
case 1 : { // viewCurrentCapacity
  msg_line1[0] = 0; msg_line2[0] = 0;
  strcat_P((char*)&msg_line1, (const char*)F("Cur: "));
  temp[0] = 0;
  floatToString((char*)&temp, uCurrent, 1);
  strcat((char*)&msg_line1, (char*)&temp);
  strcat_P((char*)&msg_line1, (const char*)F("A"));
  strcat_P((char*)&msg_line2, (const char*)F("Cap: "));
  temp[0] = 0;
  floatToString((char*)&temp, uCapacity, 0);
  strcat((char*)&msg_line2, (char*)&temp);
  strcat_P((char*)&msg_line2, (const char*)F("mAh"));
  JB.JetiBox((char*)&msg_line1, (char*)&msg_line2);
  break;
}
case 2 : { // viewVoltagePower
  msg_line1[0] = 0; msg_line2[0] = 0;
  strcat_P((char*)&msg_line1, (const char*)F("Voltage: "));
  temp[0] = 0;
  floatToString((char*)&temp, uVoltage, 2);
  strcat((char*)&msg_line1, (char*)&temp);
  strcat_P((char*)&msg_line1, (const char*)F("V"));
  strcat_P((char*)&msg_line2, (const char*)F("Power:   "));
  temp[0] = 0;
  floatToString((char*)&temp, uPower, 0);
  strcat((char*)&msg_line2, (char*)&temp);
  strcat_P((char*)&msg_line2, (const char*)F("W"));
  JB.JetiBox((char*)&msg_line1, (char*)&msg_line2);
  break;
}
case 3 : { // setSensorType
  msg_line1[0] = 0; msg_line2[0] = 0;
  strcat_P((char*)&msg_line1, (const char*)F("Set Sensor Type:"));
  if (sensorType == 0) {
    strcat_P((char*)&msg_line2, (const char*)F(" Sensor Not Set"));
  }
  if (sensorType == 1) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS712-05"));
  }
  if (sensorType == 2) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS712-20"));
  }
  if (sensorType == 3) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS712-30"));
  }
  if (sensorType == 4) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS758-50B"));
  }
  if (sensorType == 5) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS758-100B"));
  }
  if (sensorType == 6) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS758-150B"));
  }
  if (sensorType == 7) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS758-200B"));
  }
  if (sensorType == 8) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS758-50U"));
  }
  if (sensorType == 9) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS758-100U"));
  }
  if (sensorType == 10) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS758-150U"));
  }
  if (sensorType == 11) {
    strcat_P((char*)&msg_line2, (const char*)F("  ACS758-200U"));
  }
  JB.JetiBox((char*)&msg_line1, (char*)&msg_line2);
  break;
}
case 4 : { // setVoltageDivider
  msg_line1[0] = 0; msg_line2[0] = 0;
  strcat_P((char*)&msg_line1, (const char*)F("Set Max. Voltage"));
  if (voltageRange == 0) {
    strcat_P((char*)&msg_line2, (const char*)F("  Max. 14S 60V"));
  }
  if (voltageRange == 1) {
    strcat_P((char*)&msg_line2, (const char*)F("  Max. 6S 30V"));
  }
  if (voltageRange == 2) {
    strcat_P((char*)&msg_line2, (const char*)F("  Max. 3S 15V"));
  }
  if (voltageRange == 3) {
    strcat_P((char*)&msg_line2, (const char*)F("  Max. 2S/Rx 10V"));
  }
  JB.JetiBox((char*)&msg_line1, (char*)&msg_line2);
  break;
}
case 5 : { // currentCalibration
  msg_line1[0] = 0; msg_line2[0] = 0;
  strcat_P((char*)&msg_line1, (const char*)F("Current: "));
  temp[0] = 0;
  floatToString((char*)&temp, uCurrent, 2);
  strcat((char*)&msg_line1, (char*)&temp);
  strcat_P((char*)&msg_line1, (const char*)F("A"));
  strcat_P((char*)&msg_line2, (const char*)F("Calibration: "));
  temp[0] = 0;
  tempcurCalVal = curCalVal - 100;
  floatToString((char*)&temp, tempcurCalVal, 0);
  strcat((char*)&msg_line2, (char*)&temp);
  JB.JetiBox((char*)&msg_line1, (char*)&msg_line2);
  break;
}
case 6 : { // voltageCalibration
  msg_line1[0] = 0; msg_line2[0] = 0;
  strcat_P((char*)&msg_line1, (const char*)F("Voltage: "));
  temp[0] = 0;
  floatToString((char*)&temp, uVoltage, 2);
  strcat((char*)&msg_line1, (char*)&temp);
  strcat_P((char*)&msg_line1, (const char*)F("V"));
  strcat_P((char*)&msg_line2, (const char*)F("Calibration: "));
  temp[0] = 0;
  tempvoltCalVal = voltCalVal - 1000;
  floatToString((char*)&temp, tempvoltCalVal, 0);
  strcat((char*)&msg_line2, (char*)&temp);
  JB.JetiBox((char*)&msg_line1, (char*)&msg_line2);
  break;
}
case 7 : { // useReset
  msg_line1[0] = 0; msg_line2[0] = 0;
  strcat_P((char*)&msg_line1, (const char*)F("Dig.Reset: "));
  if (resetFunction == 0) {
    strcat_P((char*)&msg_line1, (const char*)F("No"));
  }
  if (resetFunction == 1) {
    strcat_P((char*)&msg_line1, (const char*)F("Yes"));
  }
  strcat_P((char*)&msg_line2, (const char*)F("Change Dn Next >"));
  JB.JetiBox((char*)&msg_line1, (char*)&msg_line2);
  break;
}
case 8 : { // storeSettings
  msg_line1[0] = 0; msg_line2[0] = 0;
  strcat_P((char*)&msg_line1, (const char*)F("Reset or Save"));
  strcat_P((char*)&msg_line2, (const char*)F("Rst UpDn Save >"));
  JB.JetiBox((char*)&msg_line1, (char*)&msg_line2);
  break;
}
case MAX_SCREEN : {
  JB.JetiBox_P(ABOUT_1, ABOUT_2);
  break;
}


Arduino/RCT-AmpSensor/Jeti_Setup.h
//
// Stuff that needs to be inside void setup()
//

/* Sensor Name
  Serial speed is 10240 for Pro Mini 3.3V due software serial
  Serial speed is 9800 for Pro Mini 5V due software serial */
JB.Init(F("RCT Amp"), JETI_RX, 9800);

// Pinmode for reset
pinMode(pinReset, INPUT_PULLUP);
// Built-in LED
pinMode(13, OUTPUT);

// Calibration settings - Reset to defaults on first use of Arduino Pro Mini
if (EEPROM.read(0) != 1) {
  EEPROM.write(0, 1); // Reset byte set to done
  EEPROM.write(1, 0); // Reset-function, not in use = 0, in use = 1
  EEPROM.write(2, 0); // Sensor type, default is Not Set
  EEPROM.write(3, 0); // Voltage range, default is 14S 60V
  EEPROM.write(4, 100); // Calibration value for current, default is 100, displayed as 0 in Jetibox
  EEPROM.put(10, 1000); // Calibration value for current, default is 1000, displayed as 0 in Jetibox
  writeEEPROM(20, 0);  // Saved capacity (mAh)
}

// Read settings
resetFunction = EEPROM.read(1);
sensorType = EEPROM.read(2);
voltageRange = EEPROM.read(3);
curCalVal = EEPROM.read(4);
EEPROM.get(10, voltCalVal);
if (resetFunction == 1) {
  uCapacity = readEEPROM(20);
  if(uCapacity > 0){
    digitalWrite(13, HIGH);   
    delay(500);              
    digitalWrite(13, LOW);
    delay(500);
    digitalWrite(13, HIGH);   
    delay(500);              
    digitalWrite(13, LOW);
    delay(500);
    digitalWrite(13, HIGH);   
    delay(500);              
    digitalWrite(13, LOW);
    }
}

// Running values
mVperAmp = mVperAmpArray[sensorType];
ACSoffset = ACSoffsetArray[sensorType];
dividerR1 = R1[voltageRange];
dividerR2 = R2[voltageRange];
tempcurCalVal = curCalVal - 100;
tempvoltCalVal = voltCalVal - 1000;

// Calculate voltage divider
divider = (float)dividerR2 / (dividerR1 + dividerR2);

/*
  Define Jeti Sensor value name & unit & variable & precision (decimals)
  JB.setValue30(JB.addData(F("NAME"), F("UNIT")), &VARIABLE, DEC);

  Sensor-value ranges:
  JB.setValue6 =   -31 to 31               (Variable is uint8_t)
  JB.setValue14 =  -8191 to 8191           (Variable is int)
  JB.setValue30 =  -536870911 to 536870911 (Variable is long)
*/

JB.setValue30(JB.addData(F("Current"), F("A")), &jetiCurrent, 1);
JB.setValue30(JB.addData(F("Capacity"), F("mAh")), &jetiCapacity, 0);
JB.setValue30(JB.addData(F("Voltage"), F("V")), &jetiVoltage, 2);
JB.setValue30(JB.addData(F("Power"), F("W")), &uPower, 0);
